### ISSUE TITLE
How to use this framework to solve steiner tree problem? 

### ISSUE THREAD

**Issue by YuLong4** [ml-jku/DIffUCO#7]
Hello, I want to use your diffusion model framework to solve the Steiner tree problem, which parts should I modify?
From Lucas(2014), I find the Ising model about the Steiner tree, 
<img width="510" alt="截屏2024-11-20 17 30 41" src="https://github.com/user-attachments/assets/85af5890-01d4-4f1e-8fd2-457623b06bb0">
<img width="285" alt="截屏2024-11-20 17 31 23" src="https://github.com/user-attachments/assets/98e43224-47fd-4a8f-b243-3e971557077b">

but the situation seems to have become very complicated. Could your framework be extended to solve the Steiner tree problem?

Looking forward to your response.

**Comment by sanokows** [comment]
good question. In principle, it would be possible to implement Steiner trees into this framework but it is indeed hard.
1. The Steiner Tree would correspond to a highly connected graph, i.e. using Graph Neural Networks is not feasible and it would be better to use a transformer architecture.
2. I have already experimented with TSP (Hamiltonian Cycles) which looks to me to have a similar structure. The problem is that if you use the TSP energy function without any inductive biases, such as softmax outputs this does not work well. The problem with softmax outputs is that those discrete diffusion models are empirically known not to work well when the categorical dimension is very high.

So my conclusion is that it won't be easy to implement and it probably will not be easy to make it work well.
If you nevertheless want to implement it you could dig a bit into the TSP code that I have written and try to understand what I have done there. But I have not experimented on TSP for a long time so I do not know if everything will run without errors.

I hope this helps you.

**Comment by sanokows** [comment]
closed due to inactivity
